api/src/main/java/io/grpc/ClientCallTracer.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Traces {@link ClientCall} events.
+ *
+ * <p>Implementers use this class to trace events on {@link ClientCall} layer. As it extends {@link
+ * ClientStreamTracer.Factory}, implementers can supply a {@link ClientStreamTracer} for tracing
+ * events on client stream layer as well.
+ *
+ * <p>Similar functionality can be achieved with a {@link ClientInterceptor} that intercepts the
+ * {@link Channel} by installing a {@link ClientStreamTracer.Factory} to {@link ClientCall}s it
+ * creates. But this mechanism allows implementers to expose direct access to tracing attributes to
+ * external components, such as {@link ClientInterceptor}s.
+ *
+ * @since 1.24.0
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/6080")
+@ThreadSafe
+public abstract class ClientCallTracer extends ClientStreamTracer.Factory {
+
+  /**
+   * Put tracer specific information as attributes to the provided builder. Tracer attributes are
+   * implementation specific.
+   *
+   * @param builder receiver of tracer information.
+   */
+  public void getTracerAttributes(Attributes.Builder builder) {}
+
+  /**
+   * A request message has been sent to the server. May be called zero or more times depending on
+   * how many messages the server is willing to accept for the operation.
+   *
+   * <p>This is called in {@link ClientCall#sendMessage(Object)}.
+   *
+   * @param message to be sent to the server.
+   */
+  public void outboundMessage(Object message) {}
+
+  /**
+   * A response message has been received. May be called zero or more times depending on whether the
+   * call response is empty, a single message or a stream of messages.
+   *
+   * <p>This is called in {@link ClientCall.Listener#onMessage(Object)}, which is invoked in call
+   * executor.
+   *
+   * @param message returned by the server
+   */
+  public void inboundMessage(Object message) {}
+
+  /**
+   * The {@link ClientCall} has been closed.
+   *
+   * <p>This is called in {@link ClientCall.Listener#onClose(Status, Metadata)}, which is invoked in
+   * call executor.
+   *
+   * @param status the result of the remote call.
+   */
+  public void callEnded(Status status) {}
```

Comment:
By looking at these tracking methods, I realize that they essentially overlap with the functionality of ClientInterceptor. I question myself whether it's really necessary to have such a large API.

Alternatively, `ClientCallTracer` could have only `getTracerAttributes()`. We don't need the new setter on `ManagedChannelBuilder`, instead we add `withClientCallTracer()` on `CallOptions` (possibly deprecate `withStreamTracerFactory()`. Keep using `ClientInterceptor` in `CensusTracingModule`, and let the interceptor to set `ClientCallTracer` for new calls, and grab the `Span` from `ClientCall.getTracerAttributes()` right after creating the call.

That looks like a much smaller change. Also ClientInterceptor probably has less impact on performance (because you don't iterate on the list of tracers on each callback). Sorry for @voidzcy that I came up this idea only after you made the code changes.

Reply:
That's ok, no worries. 
But, does that solve ordering constraint of `ClientInterceptor`s (assuming in the future we move Census modules out of core and no longer enforce Census interceptors run last)? 

Framework interceptor needs to check two places to get the `Span`. One is from `ClientCall#getTracerAttributes()` for the case when Census interceptor runs first, and the other is from `callOptions` passed to `ClientInterceptor#interceptCall()` for the case when framework interceptor runs first.

Reply:
Framework interceptors would get the `Span` only from `ClientCall#getTracerAttributes()`. As soon as the Framework interceptors get a `ClientCall` from `next.newCall(...)`, it can be sure the `ClientCallTracer` from the Census interceptor has been passed to the `ClientCallImpl`, and `getTracerAttributes()` can be immediately called, no matter which interceptor is run first.

Reply:
Oh, you are right. `next.newCall(...)` runs all the way down to `ClientCallImpl`. The `Span` will be there.

Reply:
I will have a separate PR to make the changes.